### PR TITLE
(SIMP-4186) ssh: Use features when on IPA domain

### PR DESCRIPTION
**Pull request by jeefberkey**
GSSAPIAuthentication should be disabled on the server when the host is
on an IPA domain. This commit adds a check using the new `ipa` fact to
determine whether to set GSSAPIAuthentication or not.

SIMP-4186 #comment Code changed
SIMP-4298 #close

### CODE REVIEW

**Comment by jeefberkey** [comment]
✔️  https://gitlab.com/simp/pupmod-simp-ssh/pipelines/16388217

**Comment by jeefberkey** [comment]
✔️ https://gitlab.com/simp/pupmod-simp-ssh/pipelines/16464249

File: spec/functions/ssh/global_known_hosts_spec.rb [hunk 1]
```diff
@@ -37,19 +37,19 @@
...
-            :type => 'ssh-rsa',
+            :type         => 'ssh-rsa',
             :host_aliases => facts[:fqdn].split('.').first,
-            :key => facts[:sshrsakey],
-            :ensure => 'present'
+            :key          => facts[:sshrsakey],
+            :ensure       => 'present'
           }
-          expect(resource.to_hash).to eq(expected_hash)
+          expect(resource.to_hash).to include(expected_hash)
```
Comment (op-ct) [inline spec/functions/ssh/global_known_hosts_spec.rb]: :+1:

File: manifests/server/conf.pp [hunk 2]
```diff
@@ -131,7 +131,7 @@
   Boolean                          $enable_fallback_ciphers         = true,
   Variant[Boolean,Enum['delayed']] $compression                     = false,
   Ssh::Syslogfacility              $syslogfacility                  = 'AUTHPRIV',
-  Boolean                          $gssapiauthentication            = false,
+  Boolean                          $gssapiauthentication            = $::ssh::server::params::gssapiauthentication,
```
Comment (op-ct) [inline manifests/server/conf.pp]: _If_ the SSH client should default to the same GSSAPI configuration as the SSH server, then this should refer to a new parameter in `init.pp` (which defaults to `$::ssh::server::params::gssapiauthentication`)

File: manifests/server/conf.pp [hunk 3]
```diff
@@ -161,10 +165,10 @@
     include '::haveged'
   }
 
+  $rhel_greater_then_6 = ( $facts['os']['name'] in ['RedHat','CentOS','Fedora'] ) and ( $facts['os']['release']['major'] > '6' )
```
Comment (lnemsick-simp) [inline manifests/server/conf.pp]: The logic in client/params.pp checks if the Fedora version is > 22.  So a couple of questions about this:
* Should the logic be the same
* Why are we referencing Fedora when the metadata.json does not list Fedora in list of supported OSs
* Why not use simplib::assert_metadata( $module_name )

File: CHANGELOG [hunk 4]
```diff
@@ -1,3 +1,7 @@
+* Fri Jan 19 2018 Nick Miller <nick.miller@onyxpoint.com> - 6.2.2-0
```
Comment (lnemsick-simp) [inline CHANGELOG]: I think this may warrant a minor version bump, as it is not a bug fix.

File: spec/classes/server/conf_spec.rb [hunk 5]
```diff
@@ -190,6 +192,8 @@
 
         context 'with useprivilegeseparation' do
           let(:facts) { os_facts.merge( { :openssh_version => '6.6' } ) }
+          let(:pre_condition){ "service {'sshd':}" }
```
Comment (lnemsick-simp) [inline spec/classes/server/conf_spec.rb]: Why was this added?

File: spec/classes/server/conf_spec.rb [hunk 6]
```diff
@@ -249,6 +253,17 @@
           it { is_expected.to contain_class('haveged') }
         end
 
+        context 'when connected to an IPA domain' do
+          let(:pre_condition){ 'include "::ssh"' }
+          let(:facts) {
+            os_facts.merge(
+              ipa: { 'status' => 'joined' }
+            )
```
Comment (lnemsick-simp) [inline spec/classes/server/conf_spec.rb]: Once the ipa fact is fixed (PR to be merged), 'status' is no longer a valid key.  So, you may want to either have an empty hash for ipa (you don't care what is in the hash) or have the  'domain', and 'server' keys as you did in host_config_entry_spec.rb

File: manifests/server/conf.pp (hunk 2 above)
Comment (jeefberkey) [reply manifests/server/conf.pp]: I decided against putting the logic in init.pp to fit in with the architecture of the module as it exists today.

File: manifests/server/conf.pp (hunk 3 above)
Comment (jeefberkey) [reply manifests/server/conf.pp]: 1. The logic should be the same, I can fix that
2. I don't mind Fedora in the logic, but we can't test it in this ticket. Would removing it be an issue?

File: spec/classes/server/conf_spec.rb (hunk 5 above)
Comment (jeefberkey) [reply spec/classes/server/conf_spec.rb]: The test as it was was failing on Puppet 5

File: manifests/server/conf.pp [hunk 7]
```diff
@@ -41,8 +41,9 @@
 # @param syslogfacility Gives the facility code that is used when
 #   logging messages.
 #
-# @param gssapiauthentication  Specifies whether user authentication
-#   based on GSSAPI is allowed.
+# @param gssapiauthentication Specifies whether user authentication
+#   based on GSSAPI is allowed. If the system is connected to an IPA domain,
+#   this will be set to true, based on the existance of the `ipa` fact.
```
Comment (op-ct) [inline manifests/server/conf.pp]: When the `ipa` fact is present, this parameter will default to `true`, but _can_ be overridden (using APL).

File: manifests/client/host_config_entry.pp [hunk 8]
```diff
@@ -89,7 +89,8 @@
 #   the global host key database.
 #
 # @param gssapiauthentication  Specifies whether user authentication
-#   based on GSSAPI is allowed.
+#   based on GSSAPI is allowed. If the system is connected to an IPA domain,
+#   this will be set to true, based on the existance of the `ipa` fact.
```
Comment (op-ct) [inline manifests/client/host_config_entry.pp]: When the `ipa` fact is present, the module will behave _as if_ this parameter had been set to `true`―regardless of the parameter's value (this behavior cannot be overridden).

File: manifests/server/conf.pp (hunk 3 above)
Comment (op-ct) [inline manifests/server/conf.pp]: * It will be fine to remove the Fedora logic―if it's not in the `metadata.json`, it isn't supported by SIMP, the PMT, or `on_supported_os`.
* I'd prefer to test something like `$facts['os']['family'] == 'RedHat'` instead of `['os']['name']` so it will be easier to add support for other EL distros (like Oracle/[SIMP-3620](https://simp-project.atlassian.net/browse/SIMP-3620)).
